### ISSUE TITLE
Python3.4 support

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [grpc/grpc#282]
The pure-Python code is completely compatible across 2 and 3 but it looks like the C extension code will have to be adjusted according to https://docs.python.org/3.3/howto/cporting.html.

**Comment by alex** [comment]
A good approach for PyPy would be to use the C library, but isntead of writing a C extension, bind to it using cffi (this actually can work for all versions of python)

**Comment by kryzthov** [comment]
Here's a very minimal attempt that got both examples in grpc-common working with Python 3.4.2:
https://github.com/kryzthov/grpc/commit/b7945d977a2b63965532575be820be996286924c

**Comment by nathanielmanistaatgoogle** [comment]
@kryzthov: Impressive! Will you be making that into a pull request?

**Comment by methane** [comment]
@kryzthov Ping?

**Comment by soltanmm** [comment]
This will be a post-beta thing, possibly alongside Cython-usage.

**Comment by jedsmith** [comment]
@soltanmm Just to be clear, you're saying Python support will not include 3.x until after beta?

**Comment by soltanmm** [comment]
@jedsmith To the extent of my understanding of the situation, yes.

**Comment by jedsmith** [comment]
I have to say, that's a rather surprising placement on the priority pole. Is there something I'm missing?

That's a dealbreaker for grpc on my current work (and that's okay, I'm not here to gripe, things don't always work out), but given that Python is one of the blessed at Google I'd have expected more awareness of its ecosystem and a little better support for the language, even pre-GA. Is this a proto3 limitation or a grpc limitation? Is there any way I, or anyone more familiar with Python than me, can help?

**Comment by soltanmm** [comment]
I don't believe Cython was an overreach, is an overreach, or will be an overreach. The current reasoning is that we know Cython will effectively fix the vast majority of Python 2->3 compatibility issues (the rest being small syntactic changes throughout the Python code). We know this because it was demonstrated earlier in #2243. Whether or not the rest of the team bites on Cython is partially dependent, currently, on the (relative) stability of the core API. That will be much less of an issue after beta hits. In the meanwhile, I'd rather avoid expending effort on new-target code I intend on scrapping in a short period of time (as far as the current beta timeline is concerned) while we still have a working library for a significant fraction of our targets.

**Comment by drbo** [comment]
I would also like to see PyPy support, due to its speed.  CFFI is probably the best way to go.  CFFI works on CPython, too, though there it is probably slower than a C extension module or Cython.

**Comment by leifurhauks** [comment]
I notice that the tests are not currently python3 compatible. Should that be a separate issue?

**Comment by leifurhauks** [comment]
How much work is still needed for python 3 compatibility? I might be able to contribute.

**Comment by soltanmm** [comment]
@leifurhauks I think the ideal is for everything to run from under the same roof, and I think we consider the tests as part of that notion of everything. There *shouldn't* be much needed for Python 3 compatibility, but I vaguely recall bumping my head on something when I attempted to take a quick pass at it. Something about protobuf messages somehow not getting turned into bytes by the time they reached the Cython layers. If you want to take a stab at it, or even if you just want to take out some of the low-lying fruit like Py2 -> Py3 renames and such in either the `test` or `grpc` packages, we'd really appreciate it!

**Comment by leifurhauks** [comment]
Sounds good, I'll see what I can do!

**Comment by goldenbull** [comment]
Hi everyone, any progress on python3.x compatibility?

**Comment by nathanielmanistaatgoogle** [comment]
@goldenbull: I think we're now at the stage of _some_ parts working for _some_ developers on _some_ particular machines. So... still a ways from formal support, but more than nothing.

**Comment by lovfu** [comment]
I change the function server_options and server in src/python/grpcio/grpc/beta/implementations.py as follow:

def server_options(
    multi_method_implementation=None, request_deserializers=None,
    response_serializers=None, thread_pool=None, thread_pool_size=None,
    default_timeout=None, maximum_timeout=None):

  if multi_method_implementation is not None:
      deserializers = {}
      for (k,v) in multi_method_implementation.items():
          deserializers[(k[0].encode(), k[1].encode())] = v
      multi_method_implementation.clear()
      multi_method_implementation.update(deserializers)

  if request_deserializers is not None:
      deserializers = {}
      for (k,v) in request_deserializers.items():
          deserializers[(k[0].encode(), k[1].encode())] = v
      request_deserializers.clear()
      request_deserializers.update(deserializers)

  if response_serializers is not None:
      deserializers = {}
      for (k,v) in response_serializers.items():
          deserializers[(k[0].encode(), k[1].encode())] = v
      response_serializers.clear()
      response_serializers.update(deserializers)

  return ServerOptions(
      multi_method_implementation, request_deserializers, response_serializers,
      thread_pool, thread_pool_size, default_timeout, maximum_timeout)


def server(service_implementations, options=None):
  effective_options = _EMPTY_SERVER_OPTIONS if options is None else options

  if service_implementations is not None:
      deserializers = {}
      for (k,v) in service_implementations.items():
          deserializers[(k[0].encode(), k[1].encode())] = v
      service_implementations.clear()
      service_implementations.update(deserializers)

  return _server.server(
      service_implementations, effective_options.multi_method_implementation,
      effective_options.request_deserializers,
      effective_options.response_serializers, effective_options.thread_pool,
      effective_options.thread_pool_size, effective_options.default_timeout,
      effective_options.maximum_timeout)

the type of ticket.group and ticket.method in python3.4 is bytes
the type of implementations in server is str

so i change the type of implementations in server to bytes.
then the tests of helloworld and route_guide are currently python3 compatible in linux

**Comment by viksit** [comment]
@soltanmm thanks for the work on this - it appears this support was due for a milestone on the 21st of May, 2016. Would you update on where this is now?

Thanks!

**Comment by soltanmm** [comment]
Thanks goes to @leifurhauks and @kpayson64. I believe we'll have *at least* unofficial Python 3 support starting with the next minor release (as in, it'll supposedly work but won't necessarily be tested across all platforms or it'll have some other arbitrary caveat amounting to, "your mileage may vary").

**Comment by nathanielmanistaatgoogle** [comment]
Hey everyone following this issue with interest: we're now at a point where we'd love to hear from you about how gRPC Python (at head of source control master) fails for you when run with Python 3. Please file new issues in this issue tracker with your platform, version of Python 3, and if possible running code that demonstrates the problem.

**Comment by jgeewax** [comment]
/cc @daspecster @tseaver @dhermes

On Tue, Jun 14, 2016 at 5:01 PM, Nathaniel Manista <notifications@github.com

**Comment by tseaver** [comment]
Here is my attempt to use `grpcio` 0.14.0 (plus the GAX package) in `gcloud`.  First, installing the gRPC-related packages into the `py34` Tox environment:

```bash
$ .tox/py34 bin/pip install gcloud[grpc]
...
Successfully installed enum34-1.1.6 futures-3.0.5 gax-google-pubsub-v1-0.7.7 google-gax-0.11.0 grpc-google-pubsub-v1-0.7.6 grpcio-0.14.0 ply-3.8
```

Next, running the unit tests under Python 3.4 w/ gRPC installed:

```bash
$ .tox/py34/bin/nosetests --exclude-dir=system_tests --nocapture
...........................................................................................................................................................................................................................................S..SS....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................SSSSSSSSSS..........................................................................................................FFE.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
======================================================================
ERROR: test_topic_publish_miss_w_attrs_w_bytes_payload (gcloud.pubsub.test__gax.Test_PublisherAPI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tseaver/projects/agendaless/Google/src/gcloud-python/gcloud/pubsub/test__gax.py", line 223, in test_topic_publish_miss_w_attrs_w_bytes_payload
    B64 = base64.b64encode(PAYLOAD)
  File "/home/tseaver/projects/agendaless/Google/src/gcloud-python/.tox/py34/lib/python3.4/base64.py", line 62, in b64encode
    encoded = binascii.b2a_base64(s)[:-1]
TypeError: 'str' does not support the buffer interface

======================================================================
FAIL: test_topic_publish_error (gcloud.pubsub.test__gax.Test_PublisherAPI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tseaver/projects/agendaless/Google/src/gcloud-python/gcloud/pubsub/test__gax.py", line 253, in test_topic_publish_error
    self.assertEqual(message_pb.data, B64)
AssertionError: b'VGhpcyBpcyB0aGUgbWVzc2FnZSB0ZXh0' != 'VGhpcyBpcyB0aGUgbWVzc2FnZSB0ZXh0'

======================================================================
FAIL: test_topic_publish_hit (gcloud.pubsub.test__gax.Test_PublisherAPI)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tseaver/projects/agendaless/Google/src/gcloud-python/gcloud/pubsub/test__gax.py", line 215, in test_topic_publish_hit
    self.assertEqual(message_pb.data, B64)
AssertionError: b'VGhpcyBpcyB0aGUgbWVzc2FnZSB0ZXh0' != 'VGhpcyBpcyB0aGUgbWVzc2FnZSB0ZXh0'

----------------------------------------------------------------------
Ran 2048 tests in 1.312s

FAILED (SKIP=13, errors=1, failures=2)
```

Those failures are issues in gcloud's own test suite, in testcases which haven't been exercised under Py3k until just now.

Next, I tried running the snippets for Pubsub:

```bash
$ .tox/py34/bin/python docs/pubsub_snippets.py
client_list_topics       : List topics for a project.
  ERROR: TypeError("the JSON object must be str, not 'bytes'",)
client_list_subscriptions: List all subscriptions for a project.
  ERROR: TypeError("the JSON object must be str, not 'bytes'",)
topic_create             : Create a topic.
  ERROR: TypeError("the JSON object must be str, not 'bytes'",)
topic_exists             : Test existence of a topic.
  ERROR: TypeError("the JSON object must be str, not 'bytes'",)
Traceback (most recent call last):
  File "docs/pubsub_snippets.py", line 446, in <module>
    main()
  File "docs/pubsub_snippets.py", line 443, in main
    item.delete()
  File "/home/tseaver/projects/agendaless/Google/src/gcloud-python/.tox/py34/lib/python3.4/site-packages/gcloud/pubsub/topic.py", line 196, in delete
    api = client.publisher_api
  File "/home/tseaver/projects/agendaless/Google/src/gcloud-python/.tox/py34/lib/python3.4/site-packages/gcloud/pubsub/client.py", line 74, in publisher_api
    generated = GeneratedPublisherAPI()
  File "/home/tseaver/projects/agendaless/Google/src/gcloud-python/.tox/py34/lib/python3.4/site-packages/google/pubsub/v1/publisher_api.py", line 175, in __init__
    __name__, 'publisher_client_config.json'))
  File "/opt/Python-3.4.3/lib/python3.4/json/__init__.py", line 312, in loads
    s.__class__.__name__))
TypeError: the JSON object must be str, not 'bytes'
```

That issue is inside the `gax-google-pubsub-v1` library, not `grpcio`.

**Comment by kpayson64** [comment]
@tseaver 
To clarify, the grpcio test suite passes with Python3.4 in our master branch.  The 0.14 release is missing a couple of changes that will be available in the 0.15 release shortly.

**Comment by ctiller** [comment]
@kpayson64 - if the test suite is passing, why is this still open?

**Comment by kpayson64** [comment]
There is still some ongoing work for continuously running Python3.5 and Python3.6 unit tests (We currently only have Python3.4).  I'd also like to get some Python3 performance numbers.

That being said, I think it makes sense to close this and open new issues for the above.

**Comment by ictxiangxin** [comment]
> Windows 10
> python 3.5.1
> grpcio 0.15.0

My server is simple:
`import test_pb2


class Hello(test_pb2.BetahelloworldServiceServicer):
    def hello(self, request, context):
        return test_pb2.testResponse(world="ict")

s = test_pb2.beta_create_helloworldService_server(Hello())
s.add_insecure_port('[::]:50051')
s.start()
`

And I got error:
`C:\Users\ICTxi\AppData\Local\Programs\Python\Python35\python.exe C:/Users/ICTxi/PycharmProjects/grpc-test/test_pb2.py
Traceback (most recent call last):
  File "C:/Users/ICTxi/PycharmProjects/grpc-test/test_pb2.py", line 108, in <module>
    from grpc.beta import implementations as beta_implementations
  File "C:\Users\ICTxi\AppData\Local\Programs\Python\Python35\lib\site-packages\grpc\__init__.py", line 37, in <module>
    from grpc._cython import cygrpc as _cygrpc
  File "src/python/grpcio/grpc/_cython/cygrpc.pyx", line 73, in init grpc._cython.cygrpc (./src/python/grpcio\grpc/_cython/cygrpc.c:28136)
  File "src/python/grpcio/grpc/_cython/cygrpc.pyx", line 59, in grpc._cython.cygrpc._ModuleState.__cinit__ (./src/python/grpcio\grpc/_cython/cygrpc.c:22020)
TypeError: expected bytes, str found`

**Comment by kpayson64** [comment]
@ictxiangxin

Thanks for the report.  I've opened an issue for tracking:
https://github.com/grpc/grpc/issues/7115

This should be fixed in the 0.15.1 release.
